ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v2/ReadService.scala

```diff
@@ -0,0 +1,175 @@
+    *   for the same [[SubmitterInfo.changeId]] with offset at least `off`
+    *   whose [[SubmitterInfo.submissionRank]] is at most `rank`.
+    *
+    *   If the [[WriteService]] detects that a command submission would violate the submission rank guarantee
+    *   if accepted or rejected, it either returns a [[SubmissionResult.SynchronousError]] error or
+    *   produces a [[Update.CommandRejected]] with [[Update.CommandRejected.cancelled]].
+    *
+    * - *finality*: If the corresponding [[WriteService]] acknowledges a submitted transaction or rejection
+    *   with [[SubmissionResult.Acknowledged]], the [[ReadService]] SHOULD make sure that
+    *   it eventually produces a [[Update.TransactionAccepted]] or [[Update.CommandRejected]] with this [[SubmitterInfo]],
+    *   even if there are crashes or lost network messages.
+    *
+    * The second class of properties relates multiple calls to
+    * [[stateUpdates]]s, and thereby provides constraints on which [[Update]]s
+    * need to be persisted. Before explaining them in detail we provide
+    * intuition.
+    *
+    * All [[Update]]s other than [[Update.CommandRejected]] must
+    * always be persisted by the backends implementing the [[ReadService]].
+    * For rejections, the situation is more nuanced, as we want to provide
+    * the backends with additional implementation leeway.
+    *
+    * TODO(v2) Do we actually exploit this freedom anywhere? With later [[Update.CommandRejected]]s referring
+    *  to earlier completion offsets (e.g., ALREADY_EXISTS), this becomes tricky to specify.
+    *
+    * [[Update.CommandRejected]] messages are advisory messages to submitters of
+    * transactions to inform them in a timely fashion that their transaction
+    * has been rejected.
+    *
+    * Given this intuition for the desired mechanism, we advise participant
+    * state implementations to aim to always provide timely
+    * [[Update.CommandRejected]] messages.
+    *
+    * Implementations are free to not persist [[Update.CommandRejected]] updates
+    * provided their [[Offset]]s are not reused. This is relevant for the case
+    * where a consumer rebuilds his view of the state by starting from a fresh
+    * call to [[ReadService.stateUpdates]]; e.g., because it or the
+    * stream provider crashed.
+    *
+    * Formally, we capture the expected relation between two calls
+    * `s1 = stateUpdates(o1)` and `s2 = stateUpdates(o2)` for `o1 <= o2` as
+    * follows.
+    *
+    * - *unique offsets*: for any update `u1` with offset `uo` in `s1` and any
+    *   update `u2` with the same offset `uo` in `se2` it holds that `u1 == u2`.
+    *   This means that offsets can never be reused. Together with
+    *   *strictly increasing [[Offset]]* this also implies that the order of
+    *   elements present in both `s1` and `s2` cannot change.
+    *
+    * - *persistent updates*: any update other than
+    *   [[Update.CommandRejected]] in `s2` must also be present in `s1`.
```

Comment:
This whole specification is too strong; and for example not satisfied by `sandbox-on-x`, which uses an in-memory bridge. What we need to specify is that the streams are resumable in the following sense: the single stream properties are also satisfied when consuming updates from offsets `o1` to `o3`, and then restarting from `o2` while ingoring all `Update`s in the range `[o2, o3]` received on the first stream.

Do you see this the same @andreaslochbihler-da ?

Reply:
The whole resumption requirements are taken over literally from what we currently have in `v1.ReadService`. So are you saying that our own implemenations do not conform to the requirements we currently have? Let's discuss this in person.